cpp/src/connect/utils.h

```diff
@@ -0,0 +1,240 @@
+    throw MLCommon::Exception(msg);                                            \
+  } while (0)
+
+/** macro to check for a conditional and assert on failure */
+#define ASSERT(check, fmt, ...)                                                \
+  do {                                                                         \
+    if (!(check))                                                              \
+      THROW(fmt, ##__VA_ARGS__);                                               \
+  } while (0)
+
+/** check for cuda runtime API errors and assert accordingly */
+#define CUDA_CHECK(call)                                                       \
+  do {                                                                         \
+    cudaError_t status = call;                                                 \
+    ASSERT(status == cudaSuccess, "FAIL: call='%s'. Reason:%s\n", #call,       \
+           cudaGetErrorString(status));                                        \
+  } while (0)
+
+
+
+///@todo: add a similar CUDA_CHECK_NO_THROW
+/// (Ref: https://github.com/rapidsai/cuml/issues/229)
+
+/**
+ * @brief Generic copy method for all kinds of transfers
+ * @tparam Type data type
+ * @param dst destination pointer
+ * @param src source pointer
+ * @param len lenth of the src/dst buffers in terms of number of elements
+ * @param stream cuda stream
+ */
+template <typename Type>
+void copy(Type *dst, const Type *src, size_t len, cudaStream_t stream) {
+    CUDA_CHECK(cudaMemcpyAsync(dst, src, len * sizeof(Type),
+                               cudaMemcpyDefault, stream));
+}
+
+/**
+ * @defgroup Copy Copy methods
+ * These are here along with the generic 'copy' method in order to improve
+ * code readability using explicitly specified function names
+ * @{
+ */
+/** performs a host to device copy */
+template <typename Type>
+void updateDevice(Type *dPtr, const Type *hPtr, size_t len,
+                  cudaStream_t stream) {
+  copy(dPtr, hPtr, len, stream);
+}
+
+/** performs a device to host copy */
+template <typename Type>
+void updateHost(Type *hPtr, const Type *dPtr, size_t len,
+                cudaStream_t stream) {
+  copy(hPtr, dPtr, len, stream);
+}
+
+template <typename Type>
+void copyAsync(Type* dPtr1, const Type* dPtr2, size_t len,
+               cudaStream_t stream) {
+  CUDA_CHECK(cudaMemcpyAsync(dPtr1, dPtr2, len * sizeof(Type),
+                             cudaMemcpyDeviceToDevice, stream));
+}
+/** @} */
+
+/** Helper function to calculate need memory for allocate to store dense matrix.
+* @param rows number of rows in matrix
+* @param columns number of columns in matrix
+* @return need number of items to allocate via allocate()
+* @sa allocate()
+*/
+inline size_t allocLengthForMatrix(size_t rows, size_t columns) {
+  return rows * columns;
+}
+
+/** cuda malloc */
+template <typename Type>
+void allocate(Type *&ptr, size_t len, bool setZero = false) {
```

Comment:
This bypasses RMM, and are we using this? If not, this function needs to be deleted to avoid confusion, if yes, this function needs to go through RMM.

Reply:
Yes, I'll fix that. The function is called. So are a few `cudaFree()`.